### ISSUE TITLE
Joins support subqueries, not just table names

### ISSUE THREAD

**Issue by osdiab** [knex/documentation#78]
The docs don't clarify in the examples that the first argument in each `join` method, `table`, can actually be a subquery.

Examples here for `leftJoin`: https://stackoverflow.com/questions/48098061/knex-js-subqueries-on-mysql-left-join

I just inherited a codebase that uses this behavior, I would love it if the documentation, and correspondingly the typescript types, reflect this reality.

**Comment by osdiab** [comment]
cc @tgriesser @elhigu

**Comment by elhigu** [comment]
@osdiab Yes this is definitely valid issue and documentation should be improved to mention that one can pass also `(builder) => {}`, `QueryBuilder` and `knex.raw` as a first parameter.

Some Examples:
https://runkit.com/embed/rr17trpb18gr

Funnily enough passing QueryBuilder / callback actually is inconsistent so that QueryBuilder is not wrapped to parenthesis...

**Comment by jayarjo** [comment]
@osdiab can you use that syntax with Typescript?

**Comment by osdiab** [comment]
Hmm to be honest since this was 2 years ago I’m fuzzy on the details and not working on that project anymore so I don’t think I can be very helpful here :P

**Comment by osdiab** [comment]
That said typing this use case should be simple to do in TypeScript I believe

**Comment by osdiab** [comment]
But given that PR went in I think this should be fixed lol
